app/code/Magento/Sales/Model/Order/Invoice/Total/Tax.php

```diff
@@ -118,8 +118,13 @@ public function collect(\Magento\Sales\Model\Order\Invoice $invoice)
         $invoice->setDiscountTaxCompensationAmount($taxDiscountCompensationAmt);
         $invoice->setBaseDiscountTaxCompensationAmount($baseTaxDiscountCompensationAmt);
 
-        $invoice->setGrandTotal($invoice->getGrandTotal() + $totalTax + $totalDiscountTaxCompensation);
-        $invoice->setBaseGrandTotal($invoice->getBaseGrandTotal() + $baseTotalTax + $baseTotalDiscountTaxCompensation);
+        $grandTotal = $invoice->getGrandTotal() + $invoice->getDiscountAmount() < 0.0001
+            ? 0 : $invoice->getGrandTotal() + $totalTax + $totalDiscountTaxCompensation;
+        $baseGrandTotal = $invoice->getBaseGrandTotal() + $invoice->getBaseDiscountAmount() < 0.0001
+            ? 0 : $invoice->getBaseGrandTotal() + $baseTotalTax + $baseTotalDiscountTaxCompensation;
+
+        $invoice->setGrandTotal($grandTotal);
+        $invoice->setBaseGrandTotal($baseGrandTotal);
```

Comment:
Such a solution would just work if the Tax totals is the last to run, but depending on sort order is not great, I recommend looking at `\Magento\Sales\Model\Order\Invoice\Total\Shipping` line 41 and 42, that is where the shipping cost is bein added to the grand total.